lighthouse-core/gather/computed/dependency-graph/estimator/estimator.js

```diff
@@ -0,0 +1,284 @@
+/**
+ * @license Copyright 2017 Google Inc. All Rights Reserved.
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with the License. You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the specific language governing permissions and limitations under the License.
+ */
+'use strict';
+
+const Node = require('../node');
+const TcpConnection = require('./tcp-connection');
+
+// see https://cs.chromium.org/search/?q=kDefaultMaxNumDelayableRequestsPerClient&sq=package:chromium&type=cs
+const DEFAULT_MAXIMUM_CONCURRENT_REQUESTS = 10;
+const DEFAULT_RESPONSE_TIME = 30;
+const DEFAULT_RTT = 150;
+const DEFAULT_THROUGHPUT = 1600 * 1024; // 1.6 Mbps
+
+function groupBy(items, keyFunc) {
+  const grouped = new Map();
+  items.forEach(item => {
+    const key = keyFunc(item);
+    const group = grouped.get(key) || [];
+    group.push(item);
+    grouped.set(key, group);
+  });
+
+  return grouped;
+}
+
+class Estimator {
+  /**
+   * @param {!Node} graph
+   * @param {{rtt: number, throughput: number, defaultResponseTime: number,
+   *    maximumConcurrentRequests: number}=} options
+   */
+  constructor(graph, options) {
+    this._graph = graph;
+    this._options = Object.assign(
+      {
+        rtt: DEFAULT_RTT,
+        throughput: DEFAULT_THROUGHPUT,
+        defaultResponseTime: DEFAULT_RESPONSE_TIME,
+        maximumConcurrentRequests: DEFAULT_MAXIMUM_CONCURRENT_REQUESTS,
+      },
+      options
+    );
+
+    this._rtt = this._options.rtt;
+    this._throughput = this._options.throughput;
+    this._defaultResponseTime = this._options.defaultResponseTime;
+    this._maximumConcurrentRequests = Math.min(
+      TcpConnection.maximumSaturatedConnections(this._rtt, this._throughput),
+      this._options.maximumConcurrentRequests
+    );
+  }
+
+  /**
+   * @param {!WebInspector.NetworkRequest} record
+   * @return {number}
+   */
+  static getResponseTime(record) {
+    const timing = record._timing;
+    return (timing && timing.receiveHeadersEnd - timing.sendEnd) || Infinity;
+  }
+
+  _initializeNetworkRecords() {
+    const records = [];
+
+    this._graph.getRootNode().traverse(node => {
+      if (node.type === Node.TYPES.NETWORK) {
+        records.push(node.record);
+      }
+    });
+
+    this._networkRecords = records;
+    return records;
+  }
+
+  _initializeNetworkConnections() {
+    const connections = new Map();
+    const recordsByConnection = groupBy(
+      this._networkRecords,
+      record => record.connectionId
+    );
+
+    for (const [connectionId, records] of recordsByConnection.entries()) {
+      const isSsl = records[0].parsedURL.scheme === 'https';
```

Comment:
wss?  if that should be included we have SECURE_SCHEMES elsewhere :)

Reply:
done